Streamline install progress messages

dzuelke opened the pull request:
This PR streamlines the messages printed during package installs. It cleans up several small discrepancies and issues:

1. ensure output is the same with and without `--no-progress`;
1. ensure the message doesn't say just "Downloading" after completion with `--no-progress` by appending "(100%)" when done;
1. proper-ish grammar (there should be some kind of separator before "Downloading");
1. print all progress indicators before the operation (e.g. for symlink/mirroring), which is useful in particular if the operation fails with an exception

Current, no cache:

```ShellSession
$ COMPOSER_CACHE_DIR=/dev/null composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2) Downloading: 100%         
  - Installing monolog/monolog (1.22.0) Downloading: 100%         
  - Installing symfony/routing (v3.2.2) Downloading: 100%         
  - Installing symfony/polyfill-mbstring (v1.3.0) Downloading: 100%         
  - Installing symfony/http-foundation (v3.2.2) Downloading: 100%         
  - Installing symfony/event-dispatcher (v3.2.2) Downloading: 100%         
  - Installing symfony/debug (v3.2.2) Downloading: 100%         
  - Installing symfony/http-kernel (v3.2.2) Downloading: 100%         
  - Installing pimple/pimple (v3.0.2) Downloading: 100%         
  - Installing silex/silex (v2.0.4) Downloading: 100%         
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb) Cloning 0892abb496
…
```

Current, no cache, `--no-progress`:

```ShellSession
$ COMPOSER_CACHE_DIR=/dev/null composer install --no-progress
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2) Downloading
  - Installing monolog/monolog (1.22.0) Downloading
  - Installing symfony/routing (v3.2.2) Downloading
  - Installing symfony/polyfill-mbstring (v1.3.0) Downloading
  - Installing symfony/http-foundation (v3.2.2) Downloading
  - Installing symfony/event-dispatcher (v3.2.2) Downloading
  - Installing symfony/debug (v3.2.2) Downloading
  - Installing symfony/http-kernel (v3.2.2) Downloading
  - Installing pimple/pimple (v3.0.2) Downloading
  - Installing silex/silex (v2.0.4) Downloading
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb) Cloning 0892abb496
…
```

Current, with cache:

```ShellSession
$ composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2) Loading from cache
  - Installing monolog/monolog (1.22.0) Loading from cache
  - Installing symfony/routing (v3.2.2) Loading from cache
  - Installing symfony/polyfill-mbstring (v1.3.0) Loading from cache
  - Installing symfony/http-foundation (v3.2.2) Loading from cache
  - Installing symfony/event-dispatcher (v3.2.2) Loading from cache
  - Installing symfony/debug (v3.2.2) Loading from cache
  - Installing symfony/http-kernel (v3.2.2) Loading from cache
  - Installing pimple/pimple (v3.0.2) Loading from cache
  - Installing silex/silex (v2.0.4) Loading from cache
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb) Cloning 0892abb496 from cache
…
```

Current, with cache, `--no-progress`:

```ShellSession
$ composer install --no-progress
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2) Loading from cache
  - Installing monolog/monolog (1.22.0) Loading from cache
  - Installing symfony/routing (v3.2.2) Loading from cache
  - Installing symfony/polyfill-mbstring (v1.3.0) Loading from cache
  - Installing symfony/http-foundation (v3.2.2) Loading from cache
  - Installing symfony/event-dispatcher (v3.2.2) Loading from cache
  - Installing symfony/debug (v3.2.2) Loading from cache
  - Installing symfony/http-kernel (v3.2.2) Loading from cache
  - Installing pimple/pimple (v3.0.2) Loading from cache
  - Installing silex/silex (v2.0.4) Loading from cache
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb) Cloning 0892abb496 from cache
…
```

Proposed, no cache:

```ShellSession
$ COMPOSER_CACHE_DIR=/dev/null ~/Code/oss/composer/bin/composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2): Downloading (100%)         
  - Installing monolog/monolog (1.22.0): Downloading (100%)         
  - Installing symfony/routing (v3.2.2): Downloading (100%)         
  - Installing symfony/polyfill-mbstring (v1.3.0): Downloading (100%)         
  - Installing symfony/http-foundation (v3.2.2): Downloading (100%)         
  - Installing symfony/event-dispatcher (v3.2.2): Downloading (100%)         
  - Installing symfony/debug (v3.2.2): Downloading (100%)         
  - Installing symfony/http-kernel (v3.2.2): Downloading (100%)         
  - Installing pimple/pimple (v3.0.2): Downloading (100%)         
  - Installing silex/silex (v2.0.4): Downloading (100%)         
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb): Cloning 0892abb496
…
```

Proposed, no cache, `--no-progress`:

```ShellSession
$ COMPOSER_CACHE_DIR=/dev/null ~/Code/oss/composer/bin/composer install --no-progress
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2): Downloading (100%)
  - Installing monolog/monolog (1.22.0): Downloading (100%)
  - Installing symfony/routing (v3.2.2): Downloading (100%)
  - Installing symfony/polyfill-mbstring (v1.3.0): Downloading (100%)
  - Installing symfony/http-foundation (v3.2.2): Downloading (100%)
  - Installing symfony/event-dispatcher (v3.2.2): Downloading (100%)
  - Installing symfony/debug (v3.2.2): Downloading (100%)
  - Installing symfony/http-kernel (v3.2.2): Downloading (100%)
  - Installing pimple/pimple (v3.0.2): Downloading (100%)
  - Installing silex/silex (v2.0.4): Downloading (100%)
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb): Cloning 0892abb496
…
```

Proposed, with cache:

```ShellSession
$ ~/Code/oss/composer/bin/composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2): Loading from cache
  - Installing monolog/monolog (1.22.0): Loading from cache
  - Installing symfony/routing (v3.2.2): Loading from cache
  - Installing symfony/polyfill-mbstring (v1.3.0): Loading from cache
  - Installing symfony/http-foundation (v3.2.2): Loading from cache
  - Installing symfony/event-dispatcher (v3.2.2): Loading from cache
  - Installing symfony/debug (v3.2.2): Loading from cache
  - Installing symfony/http-kernel (v3.2.2): Loading from cache
  - Installing pimple/pimple (v3.0.2): Loading from cache
  - Installing silex/silex (v2.0.4): Loading from cache
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb): Cloning 0892abb496 from cache
…
```

Proposed, with cache, `--no-progress`:

```ShellSession
$ ~/Code/oss/composer/bin/composer install --no-progress
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 11 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2): Loading from cache
  - Installing monolog/monolog (1.22.0): Loading from cache
  - Installing symfony/routing (v3.2.2): Loading from cache
  - Installing symfony/polyfill-mbstring (v1.3.0): Loading from cache
  - Installing symfony/http-foundation (v3.2.2): Loading from cache
  - Installing symfony/event-dispatcher (v3.2.2): Loading from cache
  - Installing symfony/debug (v3.2.2): Loading from cache
  - Installing symfony/http-kernel (v3.2.2): Loading from cache
  - Installing pimple/pimple (v3.0.2): Loading from cache
  - Installing silex/silex (v2.0.4): Loading from cache
  - Installing heroku/heroku-buildpack-php (dev-gracefulstop 0892abb): Cloning 0892abb496 from cache
…
```

I initially had "`Installing silex/silex (v2.0.4) from cache`" in place, but there has to be a colon too, because of various possible combinations of updates and installs, with symlinks etc, so it's "`Installing silex/silex (v2.0.4): Loading from cache`", which is then also consistent with other messages.

Lower-casing the "downloading" etc status is not feasible, as all that stuff is designed to wrap to a newline on failure, and then it would look odd. Not grammatically correct, but better than before and not worth the hassle, I'd say.

Also pretty certain I've missed a spot or two.

Thoughts?

alcohol commented:
Why did you base it onto 1.3 though? Does not really look like a bug fix to me, so master seems more appropriate?

dzuelke commented:
Mh, well, maybe. The formatting was changed in 1.3.0, so I figured, why not clean it up in a following patch release. @Seldaek?

Seldaek commented:
IMO it's fine as bugfix, and to get it out there quicker too

stof commented:
👍

dzuelke commented:
Just found more wonkiness in the current version:

```
$ composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating dzuelke/twentyfifteen-child (1.0.0 => 1.0.1)  - Removing dzuelke/twentyfifteen-child (1.0.0)
 Symlinked from themes/twentyfifteen-child
…
```

Need to fix that. @Seldaek please don't merge this yet for 1.3.2; it can wait.

dzuelke commented:
Updated to use a colon for "from cache" too, as it can't be made consistent with symlink messages etc. Also fixed a few corner cases, not all related to my changes, as e.g. symlinking printed too much info.

Now looks proper:

```
  - Removing ramsey/array_column (1.1.3)
  - Updating johnpbloch/wordpress (4.7.2 => 4.7.1): Loading from cache
  - Updating dzuelke/twentyfifteen-child (1.0.7 => 1.0.8): Symlinking from themes/twentyfifteen-child
```

Seldaek commented:
Thanks, let's hope no big regressions occur :)
